cmake-d/UseDUnittest.cmake

```diff
@@ -7,31 +7,64 @@
+    endif()
 
-  set(_testname "${_sourcefile}")
-  set(main_unittest "${CMAKE_CURRENT_BINARY_DIR}${CMAKE_FILES_DIRECTORY}/unittest.d")
-  set(_source_file "${CMAKE_CURRENT_SOURCE_DIR}/${_sourcefile}")
+    #if(_INCLUDE_DIRECTORIES)
+        set(_TEST_INCLUDE_DIRECTORIES )
+        foreach(_INCLUDE_DIRECTORY ${_INCLUDE_DIRECTORIES})
+            if(_INCLUDE_DIRECTORY) # since commented out if doesn't work for some reason
+                list(APPEND _TEST_INCLUDE_DIRECTORIES "${CMAKE_INCLUDE_FLAG_D}${_INCLUDE_DIRECTORY}")
+            endif()
+        endforeach()
+    #endif()
 
-  #message("main_unittest ${main_unittest}")
-  #message("testname ${_testname}")
-  #message("_sourcefile ${_source_file}")
+    #if(_LINK_LIBRARIES)
+        set(_TEST_LINK_LIBRARIES )
+        foreach(_LINK_LIBRARY ${_LINK_LIBRARIES})
+            if(_LINK_LIBRARY) # since commented out if doesn't work for some reason
+                list(APPEND _TEST_LINK_LIBRARIES "${CMAKE_LINK_LIBRARY_FLAG}${_LINK_LIBRARY}")
+            endif()
+        endforeach()
+    #endif()
 
-  if(NOT EXISTS ${main_unittest})
-    file(WRITE ${main_unittest} "// Generated by UseDUnittest.cmake\nint main() { return 0; }")
-  endif()
+    if(CMAKE_LIBRARY_OUTPUT_DIRECTORY)
+        set(_TEST_LIBRARY_PATH "${CMAKE_LIBRARY_PATH_FLAG}${CMAKE_LIBRARY_OUTPUT_DIRECTORY}")
+    endif()
 
-  # TODO
-  # better way to get the included directories
-  get_directory_property(include_dirs INCLUDE_DIRECTORIES)
-  set(INCLUDES )
-  foreach(include_dir IN LISTS include_dirs)
-    list(APPEND INCLUDES "${CMAKE_INCLUDE_FLAG_D}${include_dir}")
-  endforeach()
+    add_test(NAME build_${_TEST_TARGET}
+        COMMAND ${CMAKE_D_COMPILER} ${_TEST_FLAGS} ${_TEST_INCLUDE_DIRECTORIES} ${_TEST_LIBRARY_PATH} ${_TEST_LINK_LIBRARIES} -of${CMAKE_CURRENT_BINARY_DIR}/${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/${_TEST_TARGET} ${_TEST_SOURCES} -unittest ${_TEST_MAIN})
 
-  separate_arguments(CMAKE_D_FLAGS)
+    add_test(NAME ${_TEST_TARGET}
+        COMMAND ${CMAKE_CURRENT_BINARY_DIR}/${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/${_TEST_TARGET})
+    set_tests_properties(${_TEST_TARGET} PROPERTIES ENVIRONMENT "LD_LIBRARY_PATH=${CMAKE_CURRENT_BINARY_DIR}/${CMAKE_LIBRARY_OUTPUT_DIRECTORY}")
+endmacro()
 
-  add_test(NAME ${_testname}
-  	COMMAND ${CMAKE_D_COMPILER} ${CMAKE_D_FLAGS} ${INCLUDES} ${main_unittest} -unittest -run ${_source_file})
+macro(add_unittests_for_target _TARGET)
+    set(_TEST_TARGET "test_${_TARGET}")
+    
+    get_target_property(_TARGET_TEST_SOURCES ${_TARGET} SOURCES)
+    if(_TARGET_TEST_SOURCES)    
+        get_target_property(_TARGET_INCLUDE_DIRECTORIES ${_TARGET} INCLUDE_DIRECTORIES)
+        get_target_property(_TARGET_LINK_LIBRARIES ${_TARGET} LINK_LIBRARIES)
 
+        _add_unittests(${_TEST_TARGET} "${_TARGET_TEST_SOURCES}" "${_TARGET_INCLUDE_DIRECTORIES}" "${_TARGET_LINK_LIBRARIES}")
+    else()
+        message(SEND_ERROR "cannot add unittests for a source file less target")
```

Comment:
"Cannot add unittests for target ${_TARGET} with no source files"

Reply:
hmm... at my system they do
http://picpaste.com/Screenshot_from_2017-12-04_23-14-34.png

Reply:
I'm on cmake 3.10

Reply:
sorry, I was too laconic... I meant to suggest a "better" message text.